pdf/model/signature_handler.go

```diff
@@ -0,0 +1,194 @@
+	// InitSignature sets the PdfSignature parameters.
+	InitSignature(*PdfSignature) error
+	NewDigest(sig *PdfSignature) (Hasher, error)
+	Sign(sig *PdfSignature, digest Hasher) error
+}
+
+// SignatureValidationResult defines the response from the signature validation handler.
+type SignatureValidationResult struct {
+	// List of errors when validating the signature.
+	Errors      []string
+	IsSigned    bool
+	IsVerified  bool
+	IsTrusted   bool
+	Fields      []*PdfField
+	Name        string
+	Date        PdfDate
+	Reason      string
+	Location    string
+	ContactInfo string
+
+	// TODO(gunnsth): Add more fields such as ability to access the certificate information (name, CN, etc).
+	// TODO: Also add flags to indicate whether the signature covers the entire file, or the entire portion of
+	// a revision (if incremental updates used).
+}
+
+func (v SignatureValidationResult) String() string {
+	var buf bytes.Buffer
+	buf.WriteString(fmt.Sprintf("Name: %s\n", v.Name))
+	if v.Date.year > 0 {
+		buf.WriteString(fmt.Sprintf("Date: %s\n", v.Date.ToGoTime().String()))
+	} else {
+		buf.WriteString("Date not specified\n")
+	}
+	if len(v.Reason) > 0 {
+		buf.WriteString(fmt.Sprintf("Reason: %s\n", v.Reason))
+	} else {
+		buf.WriteString("No reason specified\n")
+	}
+	if len(v.Location) > 0 {
+		buf.WriteString(fmt.Sprintf("Location: %s\n", v.Location))
+	} else {
+		buf.WriteString("Location not specified\n")
+	}
+	if len(v.ContactInfo) > 0 {
+		buf.WriteString(fmt.Sprintf("Contact Info: %s\n", v.ContactInfo))
+	} else {
+		buf.WriteString("Contact info not specified\n")
+	}
+	buf.WriteString(fmt.Sprintf("Fields: %d\n", len(v.Fields)))
+	if v.IsSigned {
+		buf.WriteString("Signed: Document is signed\n")
+	} else {
+		buf.WriteString("Signed: Not signed\n")
+	}
+	if v.IsVerified {
+		buf.WriteString("Signature validation: Is valid\n")
+	} else {
+		buf.WriteString("Signature validation: Is invalid\n")
+	}
+	if v.IsTrusted {
+		buf.WriteString("Trusted: Certificate is trusted\n")
+	} else {
+		buf.WriteString("Trusted: Untrusted certificate\n")
+	}
+
+	return buf.String()
+}
+
+// ValidateSignatures validates digital signatures in the document.
+func (r *PdfReader) ValidateSignatures(handlers []SignatureHandler) ([]SignatureValidationResult, error) {
+	if r.AcroForm == nil {
+		return nil, nil
+	}
+	if r.AcroForm.Fields == nil {
+		return nil, nil
+	}
+	type sigFieldPair struct {
+		sig     *PdfSignature
+		field   *PdfField
+		handler SignatureHandler
+	}
+	var pairs []*sigFieldPair
+
+	for _, f := range r.AcroForm.AllFields() {
+		if f.V == nil {
+			continue
+		}
+		if d, found := core.GetDict(f.V); found {
+			if name, ok := core.GetNameVal(d.Get("Type")); ok && name == "Sig" {
+				ind, _ := core.GetIndirect(f.V) // TODO refactoring
+				sig, err := r.newPdfSignatureFromIndirect(ind)
+				if err != nil {
+					return nil, err
+				}
+				pairs = append(pairs, &sigFieldPair{sig: sig, field: f})
```

Comment:
Finding the applicable handler could be done here in order to avoid extra iterations:

```
pair := &sigFieldPair{sig: sig, field: f}
for _, handler := range handlers {
    if handler.IsApplicable(pair.sig) {
        pair.handler = handler
        break
    }
}

pairs = append(pairs, pair)
```